#2530 Enable SASL for bie kafka

agile-josiah opened the pull request:
<!-- Ensure the PR title reflects the feature or bug name -->

## What was the problem?

The VRO Certs to access Kafka are expiring and we need to decide whether VRO will provision their own Certs and provide the Public Key to BIP Solutions to renew the ACL access or attempt to complete the RBAC migration making the Certs a non-issue.

Associated tickets or Slack threads:
- #2530 
- https://dsva.slack.com/archives/C04U8GESRT7/p1706727018289089

## How does this fix it?[^1]
This enables SASL_SSL for bie-kafka comms
## How to test this PR
- run source ./scripts/kafka-service.sh
- Watch the connection between mock-bie-kafka and svc-bie-kafka


[^1]: [Pull-Requests guidelines](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Pull-Requests). If PR is significant, update [Current Software State](https://github.com/department-of-veterans-affairs/abd-vro/wiki/Current-Software-State) wiki page.
[^secrel]: To check if a PR will succeed in the SecRel workflow, [test PRs in the SecRel pipeline](https://github.com/department-of-veterans-affairs/abd-vro-internal/wiki/Secure-Release-process#to-test-prs-in-the-secrel-pipeline).

agile-josiah committed:
Initial attempt currently failing on sasl handshake

agile-josiah committed:
Enables sasl_ssl for our kafka broker and clients

agile-josiah committed:
More changes

agile-josiah committed:
Merge remote-tracking branch 'origin/develop' into josiah/#2530/kafka-sasl

agile-josiah committed:
Fix for integration test

tejans24 commented on svc-bie-kafka/src/main/resources/application.yaml:
```diff
@@ -14,21 +14,28 @@ spring:
       auto-offset-reset: earliest
       # Comment out the following line to connect without SSL
-      security.protocol: SSL
       ssl:
```
is this needed if we are no longer using ssl for this?

agile-josiah replied:
We are using SASL_SSL which still needs certs because it uses TLS encryption. The authentication method is defined as PLAIN, which will use the SSL encryption protocol to transmit a username:password. So we still will need the certs if we want to use SASL_SSL. There is an alternative to use SASL_PLAINTEXT which doesn't use TLS encryption, but it seems like the BIE folks want us to use SASL_SSL.

tejans24 replied:
Ah that makes sense. Thanks for clarifying. They have hinted that this is a replacement and not an addition to auth. But would be great to clarify with them.

tejans24 commented on svc-bie-kafka/src/main/resources/application-integration-test.yaml:
```diff
@@ -6,13 +6,21 @@ spring:
+      security.inter.broker.protocol: SASL_SSL
+      sasl:
+        mechanism: PLAIN
+        username: USERNAME
+        password: PASSWORD
```
we'll need to add the password to the Vault/kube secrets and inject it into the containers per env

msnwatson commented on mocks/docker-compose.yml:
```diff
@@ -94,13 +95,22 @@ services:
+      KAFKA_CFG_SASL_MECHANISM_CONTROLLER_PROTOCOL: PLAIN
+#      KAFKA_CONTROLLER_USER: USERNAME
+#      KAFKA_CONTROLLER_PASSWORD: PASSWORD
+#      KAFKA_INTER_BROKER_USER: USERNAME
```
Can we remove these commented lines?

agile-josiah replied:
Done.

msnwatson commented on mocks/docker-compose.yml:
```diff
@@ -94,13 +95,22 @@ services:
       KAFKA_TLS_TYPE: JKS
       KAFKA_CERTIFICATE_PASSWORD: mockKafka
-
+#
```
Is this line necessary?

msnwatson commented on svc-bie-kafka/src/main/resources/application-integration-test.yaml:
```diff
@@ -21,7 +29,7 @@ spring:
         trust-store-type: "PKCS12"
     admin:
-      security.protocol: SSL
+#      security.protocol: SSL
```
I see we just removed these kinds of lines in another place.

agile-josiah replied:
Removed

msnwatson reviewed (approved):
Some comments

agile-josiah committed:
Remove comments and unnecessary duplication of ssl

agile-josiah committed:
Merge remote-tracking branch 'origin/develop' into josiah/#2530/kafka-sasl

chengjie8 committed:
Merge branch 'develop' into josiah/#2530/kafka-sasl

abd-vro-machine committed:
Automated commit: Unpin versions of updated images

dependabot[bot] committed:
Bump org.springframework.boot:spring-boot-autoconfigure from 3.2.2 to 3.2.3 (#2659)

Bump org.springframework.boot:spring-boot-autoconfigure

Bumps [org.springframework.boot:spring-boot-autoconfigure](https://github.com/spring-projects/spring-boot) from 3.2.2 to 3.2.3.
- [Release notes](https://github.com/spring-projects/spring-boot/releases)
- [Commits](https://github.com/spring-projects/spring-boot/compare/v3.2.2...v3.2.3)

---
updated-dependencies:
- dependency-name: org.springframework.boot:spring-boot-autoconfigure
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Cheng <89034346+chengjie8@users.noreply.github.com>

dfitchett committed:
EP Merge: Special Issue Code workaround for testing (#2645)

Made special issue code an environment variable, so it can be switched without code change.

msnwatson committed:
Update QA BGS URL (#2622)

agile-josiah committed:
Merge remote-tracking branch 'origin/develop' into josiah/#2530/kafka-sasl

agile-josiah committed:
Update to use vault values, added to dev and integration test yamls

agile-josiah committed:
Merge remote-tracking branch 'origin/develop' into josiah/#2530/kafka-sasl

agile-josiah committed:
Update the remaining environments

agile-josiah committed:
Merge branch 'develop' into josiah/#2530/kafka-sasl

agile-josiah committed:
Updating the names of the topics and add the tls certs needed for communication.

agile-josiah committed:
Add the var initialization to setenv.sh

nelsestu commented on shared/lib-bie-kafka/src/main/java/gov/va/vro/model/biekafka/ContentionEvent.java:
```diff
@@ -20,9 +20,9 @@ public String getTopicName() {
-    // remove first word prefix from topic seperated by _
-    String noPrefixTopic = topic.substring(topic.indexOf("_") + 1);
-
+    // TODO(3/5/24): Add a better
```
I do like the way this comment starts, but I'd prefer a little more clarification about what better something is coming.  You may already be planning to circle back with clarity on this, but what is your intent. If we are planning to leave this TODO in place, let's clarify further, or if you can address it now, please follow through with adding the better and remove the TODO.

agile-josiah replied:
Hahaha thanks for calling this out, that TODO was out of date as soon as I wrote it because @chengjie8 supplied the regex below. I didn't plan on leaving the TODO in place post-merging.

nelsestu commented on svc-bie-kafka/src/main/resources/application-dev.yaml:
```diff
@@ -1,24 +1,23 @@
 spring:
   kafka:
-    bootstrap-servers: "${BIE_KAFKA_PLACEHOLDERS_BROKERS:kafka.dev.bip.va.gov:443}"
+    bootstrap-servers: "${BIE_KAFKA_PLACEHOLDERS_BROKERS:bip-kafka.tst.bip.va.gov:443}"
```
If we are updating urls there are a probably some docs that we should plan to synchronize with this change.  I added an acceptance criteria to update docs related to the changes introduced with this PR, but I'd be just as happy to see the docs update created as its own scope of work for next sprint: #2530

agile-josiah committed:
Update urls to dev from tst

agile-josiah committed:
Merge remote-tracking branch 'origin/develop' into josiah/#2530/kafka-sasl

agile-josiah committed:
Update Contention Event to use new updated topic names

agile-josiah committed:
Merge remote-tracking branch 'origin/develop' into josiah/#2530/kafka-sasl

agile-josiah committed:
Update BieMessagePayload to use new updated topic names

agile-josiah committed:
Updating Login config to only include Kafka user and no default value

dependabot[bot] committed:
Bump org.eclipse.jgit:org.eclipse.jgit from 6.8.0.202311291450-r to 6.9.0.202403050737-r (#2725)

Bump org.eclipse.jgit:org.eclipse.jgit

Bumps org.eclipse.jgit:org.eclipse.jgit from 6.8.0.202311291450-r to 6.9.0.202403050737-r.

---
updated-dependencies:
- dependency-name: org.eclipse.jgit:org.eclipse.jgit
  dependency-type: direct:production
  update-type: version-update:semver-minor
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Ponnia <151670616+Ponnia-M@users.noreply.github.com>

dependabot[bot] committed:
Bump com.fasterxml.jackson.core:jackson-databind from 2.16.1 to 2.16.2 (#2728)

Bumps [com.fasterxml.jackson.core:jackson-databind](https://github.com/FasterXML/jackson) from 2.16.1 to 2.16.2.
- [Commits](https://github.com/FasterXML/jackson/commits)

---
updated-dependencies:
- dependency-name: com.fasterxml.jackson.core:jackson-databind
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Ponnia <151670616+Ponnia-M@users.noreply.github.com>

dependabot[bot] committed:
Bump org.apache.commons:commons-compress from 1.26.0 to 1.26.1 (#2727)

Bumps org.apache.commons:commons-compress from 1.26.0 to 1.26.1.

---
updated-dependencies:
- dependency-name: org.apache.commons:commons-compress
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Ponnia <151670616+Ponnia-M@users.noreply.github.com>

dependabot[bot] committed:
Bump com.fasterxml.jackson.datatype:jackson-datatype-jsr310 from 2.16.1 to 2.16.2 (#2726)

Bump com.fasterxml.jackson.datatype:jackson-datatype-jsr310

Bumps com.fasterxml.jackson.datatype:jackson-datatype-jsr310 from 2.16.1 to 2.16.2.

---
updated-dependencies:
- dependency-name: com.fasterxml.jackson.datatype:jackson-datatype-jsr310
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>
Co-authored-by: Ponnia <151670616+Ponnia-M@users.noreply.github.com>

dfitchett committed:
EP Merge: Updated logging of errors and warnings (#2733)

* Updated logging of errors and warnings

nanotone committed:
Simplify EP Merge integration test sorter (#2683)

Co-authored-by: Erik Nelsestuen <eriknelsestuen@navapbc.com>
Co-authored-by: Derek Fitchett <135860892+dfitchett@users.noreply.github.com>

ponny-m committed:
Github action version rollback (#2737)

* rollback version for yoomlam/delete-older-releases

* Revert "rollback version for yoomlam/delete-older-releases"

This reverts commit 19ad81ecd940dac950f477c57021ffb445f3d8cd.

* rollback github-action version

dfitchett committed:
EPMerge: Fix serialization issue resulting in bad request response from BIP API (#2732)

Changed rest template to use configured objectmapper

dfitchett committed:
mock-bip-claims-api: Updated mocks for new EE EP Merge End2End tests. (#2746)

Updated mocks for new EE EP Merge End2End tests.

dfitchett committed:
EP Merge: add checks for get EP400 claim details (#2747)

* Added checks for EP 400 claim details to only allow claims with EP codes in [400, 409] with a benefit claim type code of '400SUPP'

lukey-luke committed:
bump pydantic version for Contention Classification App (#2674)

bump pydantic version, use model validator w/ mode="before" for validating CfI

Co-authored-by: Erik Nelsestuen <eriknelsestuen@navapbc.com>

agile-josiah committed:
#2749: PROD FIX, Increase rabbitmq container memory (#2750)

Production FIX: Increase the rabbitmq container memory

agile-josiah committed:
Update SpringBootStarterWeb (#2752)

Snyk fix, update spring-boot-web to 3.2.3

chengjie8 committed:
Upgrade spring-boot-starter-web package version  (#2754)

Upgrade spring-boot-starter-web package version

msnwatson committed:
remove all versioning (#2741)

* remove all versioning

* undo some changes

* run image-names script

* add back sourcing file

agile-josiah committed:
Merge remote-tracking branch 'origin/develop' into josiah/#2530/kafka-sasl

agile-josiah committed:
Update the substring env instead of a prefix for bia kafka topics.

agile-josiah committed:
Merge remote-tracking branch 'origin/develop' into josiah/#2530/kafka-sasl

agile-josiah committed:
Merge remote-tracking branch 'origin/develop' into josiah/#2530/kafka-sasl

agile-josiah committed:
Experiment with bie-kafka-end2end-test.yml

agile-josiah committed:
Experiment with bie-kafka-end2end-test.yml

agile-josiah committed:
Experiment with bie-kafka-end2end-test.yml

agile-josiah committed:
Update group-id

agile-josiah committed:
Merge remote-tracking branch 'origin/develop' into josiah/#2530/kafka-sasl

agile-josiah committed:
Merge remote-tracking branch 'origin/develop' into josiah/#2530/kafka-sasl

agile-josiah committed:
Add V02 to the end of topic names

agile-josiah committed:
Update the escaped ""

agile-josiah committed:
Merge remote-tracking branch 'origin/develop' into josiah/#2530/kafka-sasl

agile-josiah committed:
Changes to the key and value -deserializer

agile-josiah committed:
Changes to the key and value -deserializer

agile-josiah committed:
Add deserializer delegate class

agile-josiah committed:
Add deserializer delegate class

agile-josiah committed:
Fix integration test attempt #1

agile-josiah committed:
Update kafka cert generation script to only generate the truststore.

agile-josiah committed:
Add yes back

agile-josiah committed:
Revert config

agile-josiah committed:
Update application yamls

agile-josiah committed:
Debug application integration test yaml

agile-josiah committed:
Changes to bie kafka docker vars

agile-josiah committed:
Merge remote-tracking branch 'origin/develop' into josiah/#2530/kafka-sasl

agile-josiah committed:
Changes to bie kafka docker vars

agile-josiah committed:
Changes to bie kafka docker vars

agile-josiah committed:
Add truststore back

agile-josiah committed:
Edits to ssl/sasl connection

agile-josiah committed:
Updates to docker-entryprep.sh to ignore the keystore creation/verification

agile-josiah committed:
Remove references to keystore file

agile-josiah committed:
Add a ssl piece to the properties file

agile-josiah committed:
Add ssl truststore everywhere.

agile-josiah committed:
Change shape of schema.registry.truststore.*

agile-josiah committed:
Add in USER_INFO basic auth for schema registry

agile-josiah committed:
Revert higher environments for RBAC to match develop

agile-josiah committed:
Merge remote-tracking branch 'origin/develop' into josiah/#2530/kafka-sasl

msnwatson commented on svc-bie-kafka/src/main/resources/application-dev.yaml:
```diff
@@ -1,24 +1,32 @@
+        trust-store-type: "PKCS12"
+
+bie:
+  kafka-topic-env: "TST"
```
Nit: newline at EOF

agile-josiah replied:
Done

agile-josiah committed:
Small nits and update to infix from env

agile-josiah committed:
Merge remote-tracking branch 'origin/develop' into josiah/#2530/kafka-sasl

agile-josiah committed:
rm unnecessary Kafka cert gen

agile-josiah committed:
Add qa and sandbox as well

agile-josiah committed:
Revert mock BIE kafka to existing state

agile-josiah committed:
Merge remote-tracking branch 'origin/develop' into josiah/#2530/kafka-sasl

agile-josiah committed:
Revert local config

nelsestu reviewed (approved):
The BIE team is now going to be the BIA team, and as auth is now using the BIA RBAC authentication, it simplfies some of the certificate generation and handling.  As Teja mentioned, we'll need to include the username and password in our reimplementation of secret automation.

chengjie8 committed:
RBAC user id password are not needed for local test

chengjie8 committed:
Merge branch 'develop' into josiah/#2530/kafka-sasl
